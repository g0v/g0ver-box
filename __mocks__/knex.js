import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
