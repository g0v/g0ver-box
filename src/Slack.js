import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
