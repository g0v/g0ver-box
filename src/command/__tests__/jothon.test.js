import { client } from 'knex';
import Slack from '../../Slack';
import index from '../';

describe('jothon command', () => {
  const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'jothon 基礎建設松' };

  it('when event does not exist', async () => {
    client.mockReturnValueOnce([]);
    client.mockReturnValueOnce(['5']);

    await index(data);
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入 RFC2822 或 ISO8601 格式的 活動時間 預設為 +08 時區 （ex. 12-31 23:00+08）？\n`（輸入 exit 可離開）`',
    });

    await index({ ...data, text: '12-31 14:00' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入 活動資訊 / 報名 連結？\n`（輸入 skip 可跳過）` `（輸入 exit 可離開）`',
    });

    await index({ ...data, text: 'https://events.g0v.tw/' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith(expect.objectContaining({
      channel: 'D100',
      text: '請確認資料如下？（輸入 yes 完成建立）\n`（輸入 exit 可離開）`',
    }));

    await index({ ...data, text: 'yes' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: 'Done, jothon 基礎建設松',
    });

    expect(client).toMatchSnapshot();
  });

  it('when event is existed', async () => {
    client.mockReturnValueOnce([{ id: 5 }]);

    await index(data);
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: 'Sorry! 發現相同的活動名稱',
    });

    expect(client).toMatchSnapshot();
  });

  it('when use skip', async () => {
    client.mockReturnValueOnce([]);
    await index(data);
    await index({ ...data, text: '12-31 16' });
    await index({ ...data, text: 'skip' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith(expect.objectContaining({
      channel: 'D100',
      text: '請確認資料如下？（輸入 yes 完成建立）\n`（輸入 exit 可離開）`',
    }));
    await index({ ...data, text: 'skip' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith(expect.objectContaining({
      channel: 'D100',
      text: '無法判斷的指令',
    }));
    await index({ ...data, text: 'exit' });
  });

  describe('datetime', () => {
    beforeEach(async () => {
      client.mockReturnValueOnce([]);

      await index(data);
    });

    it('12-31 16', async () => {
      await index({ ...data, text: '12-31 16' });
      await index({ ...data, text: 'https://events.g0v.tw/' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith(expect.objectContaining({
        attachments: expect.arrayContaining([expect.objectContaining({
          color: '#000',
          text: expect.stringContaining('1514707200000'),
        })]),
      }));
      await index({ ...data, text: 'exit' });
    });

    it('12-31 16+09', async () => {
      await index({ ...data, text: '12-31 16+09' });
      await index({ ...data, text: 'https://events.g0v.tw/' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith(expect.objectContaining({
        attachments: expect.arrayContaining([expect.objectContaining({
          color: '#000',
          text: expect.stringContaining('1514703600000'),
        })]),
      }));
      await index({ ...data, text: 'exit' });
    });

    it('2017-12-31 16+09', async () => {
      await index({ ...data, text: '2017-12-31 16+09' });
      await index({ ...data, text: 'https://events.g0v.tw/' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith(expect.objectContaining({
        attachments: expect.arrayContaining([expect.objectContaining({
          color: '#000',
          text: expect.stringContaining('1514703600000'),
        })]),
      }));
      await index({ ...data, text: 'exit' });
    });
  });
});
