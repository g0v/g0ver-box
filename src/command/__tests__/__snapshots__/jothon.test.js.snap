// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jothon command when event does not exist 1`] = `
Array [
  "select * from event where deleted_at is null and title = 基礎建設松 limit 1",
  "insert into event (archive, created_at, datetime, title, updated_at, user_id) values ({\\"url\\":\\"https://events.g0v.tw/\\"}, DATE, DATE, 基礎建設松, DATE, U03B2AB13) returning id",
]
`;

exports[`jothon command when event is existed 1`] = `
Array [
  "select * from event where deleted_at is null and title = 基礎建設松 limit 1",
]
`;
